src/version_selection.rs

```diff
@@ -0,0 +1,664 @@
+                continue;
+            }
+
+            let (prefix, target) = if filename.starts_with("JuliaManifest-v") {
+                ("JuliaManifest-v", &mut highest_julia)
+            } else {
+                ("Manifest-v", &mut highest_manifest)
+            };
+
+            if let Some(version) = filename
+                .strip_prefix(prefix)
+                .and_then(|s| s.strip_suffix(".toml"))
+                .and_then(parse_version_lenient)
+            {
+                update_highest(target, version, path);
+            }
+        }
+    }
+
+    // Return highest version, preferring JuliaManifest for same version
+    match (highest_julia, highest_manifest) {
+        (Some((jv, jpath)), Some((mv, mpath))) => {
+            if jv >= mv {
+                Some(jpath)
+            } else {
+                Some(mpath)
+            }
+        }
+        (Some((_, jpath)), None) => Some(jpath),
+        (None, Some((_, mpath))) => Some(mpath),
+        (None, None) => None,
+    }
+}
+
+// Parse a version string leniently, handling incomplete versions like "1.11" or "1"
+pub fn parse_version_lenient(version_str: &str) -> Option<Version> {
+    // First try parsing as-is
+    if let Ok(version) = Version::parse(version_str) {
+        return Some(version);
+    }
+
+    // If that fails, try adding missing components
+    let parts: Vec<&str> = version_str.split('.').collect();
+    let normalized = match parts.len() {
+        1 => format!("{}.0.0", parts[0]),
+        2 => format!("{}.{}.0", parts[0], parts[1]),
+        _ => return None,
+    };
+
+    Version::parse(&normalized).ok()
+}
+
+pub fn read_manifest_julia_version(path: &Path) -> Result<Option<String>> {
+    if !path.exists() {
+        log::debug!(
+            "VersionDetect::Manifest file `{}` not found while attempting to resolve Julia version.",
+            path.display()
+        );
+        return Ok(None);
+    }
+
+    let manifest_content = fs::read_to_string(path)
+        .with_context(|| format!("Failed to read manifest file `{}`.", path.display()))?;
+
+    let manifest: Value = toml::from_str(&manifest_content).with_context(|| {
+        format!(
+            "Failed to parse manifest file `{}` as TOML.",
+            path.display()
+        )
+    })?;
+
+    Ok(manifest
+        .get("julia_version")
+        .and_then(|v| v.as_str())
+        .map(|s| s.to_string()))
+}
+
+pub fn parse_db_version(version: &str) -> Result<Version> {
+    let base = version
+        .split('+')
+        .next()
+        .ok_or_else(|| anyhow!("Invalid version string `{}`.", version))?;
+    Version::parse(base).with_context(|| format!("Failed to parse version `{}`.", base))
+}
+
+pub fn max_available_version(versions_db: &JuliaupVersionDB) -> Result<Option<Version>> {
+    Ok(versions_db
+        .available_versions
+        .keys()
+        .filter_map(|key| parse_db_version(key).ok())
+        .max())
+}
```

Comment:
A bit too functional for rust imo, would be more idiomatic to write smth like

```suggestion
impl JuliaVersionDB {
    pub fn max_available_version(versions_db: &self) -> Result<Option<Version>> {
        Ok(versions_db
            .available_versions
            .keys()
            .filter_map(|key| parse_db_version(key).ok())
            .max())
    }
}
```

Then the call would be `versions_db.max_available_version()`